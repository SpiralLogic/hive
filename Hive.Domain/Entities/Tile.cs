using System.Collections.Generic;
using System.Linq;

namespace Hive.Domain.Entities
{
    public class Tile
    {
        public Tile(int id, int playerId, string content, IEnumerable<Coordinates> moves = null)
        {
            Id = id;
            PlayerId = playerId;
            Content = content;
            Moves = moves?.ToList() ?? new List<Coordinates>();
        }

        public int Id { get; }
        public int PlayerId { get; }
        public string Content { get; }
        public ICollection<Coordinates> Moves { get; }
    }
}
