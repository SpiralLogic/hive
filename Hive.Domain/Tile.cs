using System.Collections.Generic;

namespace Hive.Domain
{
    public class Tile
    {
        private int _count=0;

        public Tile(string owner, TextContent content, string color)
        {
            Id = (++_count).ToString();
            Owner = owner;
            Content = content;
            Color = color;
        }

        public string Id { get; }
        public string Owner { get; }
        public TextContent Content { get; }
        public string Color { get; }
        public ICollection<GameCoordinate> AvailableMoves { get; } = new List<GameCoordinate>();
    }
}