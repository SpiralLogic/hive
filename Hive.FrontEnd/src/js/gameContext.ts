import { createContext, useEffect, useState } from 'react';
import { Color, GameState, HexEngine, Move, PlayerId } from './domain';

export interface GameContext {
    gameState: GameState;
    moveTile: (move: Move) => void;
    getPlayerColor: (playerId: PlayerId) => Color;
}

export const Context = createContext<GameContext>({
    gameState: {
        hexagons: [],
        players: [],
    },
    moveTile: () => undefined,
    getPlayerColor: (): string => '',
});

const getPlayerColor = (playerId: PlayerId) => {
    const playerColors = ['#85dcbc', '#f64c72'];
    return playerColors[playerId] || 'red';
};

export const useGameContext = (engine: HexEngine): [boolean, GameContext] => {
    const [loading, setLoading] = useState(true);

    const [gameState, setGameState] = useState<GameState>({
        hexagons: [],
        players: [],
    });

    useEffect(() => {
        setLoading(true);
        engine.initialState().then(state => {
            setGameState(state);
            setLoading(false);
        });
    }, [engine]);

    const moveTile = (move: Move) => {
        engine.moveTile(move).then(nextState => setGameState(nextState));
    };

    return [loading, { gameState, moveTile, getPlayerColor }];
};
