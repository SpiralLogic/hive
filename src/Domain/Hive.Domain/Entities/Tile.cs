using System.Collections.Generic;

namespace Hive.Domain.Entities
{
    public sealed record Tile(int Id, int PlayerId, Creature Creature)
    {
        public ISet<Coords> Moves { get; init; } = new HashSet<Coords>();
    }
}