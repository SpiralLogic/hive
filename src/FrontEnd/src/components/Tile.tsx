import { FunctionComponent, h } from 'preact';
import { Tile } from '../domain';
import { TileAction } from '../utilities/hive-dispatcher';
import { addHiveEventListener, useClassReducer, useHiveDispatcher } from '../utilities/hooks';
import { deepEqual } from 'fast-equals';
import { handleDrop, handleKeyboardNav, isEnterOrSpace } from '../utilities/handlers';
import { memo } from 'preact/compat';
import { useEffect, useState } from 'preact/hooks';

const tileSelector = `[tabindex].tile`;
const cellSelector = `[tabindex].cell`;
const playerSelector = `[tabindex].name`;

const TileFC: FunctionComponent<Tile> = (tile: Tile) => {
  const { id, moves, creature, playerId } = tile;
  const [focus, setFocus] = useState(tileSelector);
  const [classes, setClassList] = useClassReducer(`player${playerId} hex tile`);
  const hiveDispatcher = useHiveDispatcher();

  const deselect = () => {
    if (!classes.includes('selected')) return;
    setClassList({ type: 'remove', class: 'selected' });
    hiveDispatcher.dispatch({ type: 'tileDeselected', tile: tile });
  };

  const select = () => {
    if (classes.includes('selected')) return;
    hiveDispatcher.dispatch({ type: 'tileClear' });
    setClassList({ type: 'add', class: 'selected' });
    hiveDispatcher.dispatch({ type: 'tileSelected', tile: tile });
  };

  addHiveEventListener<TileAction>('tileSelect', (event: TileAction) => {
    if (event.tile.id == id) select();
  });

  addHiveEventListener<TileAction>('tileDeselect', (event: TileAction) => {
    if (event.tile.id == id) deselect();
  });

  addHiveEventListener('tileClear', () => {
    deselect();
    setFocus('');
  });

  const handleDragStart = (event: DragEvent) => {
    event.stopPropagation();
    select();
    setClassList({ type: 'add', class: 'beforeDrag' });
    setTimeout(() => setClassList({ type: 'remove', class: 'beforeDrag' }), 1);
  };
  const handleDragEnd = () => {
    hiveDispatcher.dispatch({ type: 'tileClear' });
    hiveDispatcher.dispatch({ type: 'tileDropped', tile: tile });
  };
  const handleClick = (event: MouseEvent) => {
    event.stopPropagation();
    !classes.includes('selected') ? select() : deselect();
  };
  const handleKeyDown = (e: KeyboardEvent) => {
    if (handleKeyboardNav(e) || !isEnterOrSpace(e)) return;
    e.stopPropagation();
    if (!classes.includes('selected')) {
      select();
      setFocus(cellSelector);
    } else {
      deselect();
    }
  };
  const handleMouseLeave = (event: { currentTarget: HTMLElement }) => event.currentTarget.blur();

  useEffect(() => {
    if (!focus) return;
    const focusElement =
      document.querySelector<HTMLElement>(focus) || document.querySelector<HTMLElement>(playerSelector);
    focusElement?.focus();
    setFocus('');
  }, [focus]);

  const attributes = {
    title: creature,
    class: classes,
    draggable: !!moves.length,
    tabindex: moves.length ? 0 : undefined,
  };
  const handlers = moves.length
    ? {
        onclick: handleClick,
        ondragstart: handleDragStart,
        ondragend: handleDragEnd,
        onkeydown: handleKeyDown,
        onmouseleave: handleMouseLeave,
        ondrop: handleDrop,
      }
    : {
        ondrop: handleDrop,
      };
  return (
    <div {...attributes} {...handlers}>
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 100 100">
        <use href="#hex" />
        <use class="creature" href={`#${creature.toLowerCase()}`} />
      </svg>
    </div>
  );
};
TileFC.displayName = 'Tile';
export default memo(TileFC, deepEqual);
