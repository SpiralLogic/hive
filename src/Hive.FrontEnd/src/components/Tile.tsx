import '../css/tile.css';

import { FunctionComponent } from 'preact';

import Hexagon from './Hexagon';
import { computed, Signal } from '@preact/signals';
import { useClassSignal } from '../hooks/useClassReducer';

type Properties = {
  selected?: boolean;
  creature?: string;
  classes: Signal<string>;
  classAction: ReturnType<typeof useClassSignal>[1];

  [rest: string]: unknown;
};

const Tile: FunctionComponent<Properties> = (properties) => {
  const { classes, children, classAction, selected, creature, ...rest } = properties;
  classAction.add('tile');
  if (creature) classAction.add(`${creature.toLowerCase()}`);
  if (selected) classAction.add('selected');

  const svgs = creature ? <use className={`creature`} href={`#${creature.toLowerCase()}`} /> : undefined;
  return (
    <Hexagon svgs={svgs} {...rest} classes={classes}>
      {children}
    </Hexagon>
  );
};
Tile.displayName = 'Tile';
export default Tile;
