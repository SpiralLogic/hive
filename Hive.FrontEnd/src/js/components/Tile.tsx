import * as React from 'react';
import { HexCoordinates, PlayerId, TextContent, TileId } from '../domain';
import { tileDragEmitter } from '../gameContext';
import Cell from './Cell';

type Props = {
    id: TileId,
    content: TextContent
    playerId: PlayerId,
    availableMoves: HexCoordinates[],
} & typeof defaultProps;
const defaultProps = {
    tileDragEmitter: tileDragEmitter
};

const getPlayerColor = (playerId: PlayerId) => {
    const playerColors = ['#85dcbc', '#f64c72'];
    return playerColors[playerId] || 'red';
};

function Tile (props: Props) {
    const { id, availableMoves, content, playerId, tileDragEmitter } = props;

    function handleDragStart (ev: React.DragEvent<HTMLDivElement>) {
        tileDragEmitter.emit({ type: 'start', tileId: id, tileMoves: availableMoves });
    }

    function handleDragEnd (ev: React.DragEvent<HTMLDivElement>) {
        tileDragEmitter.emit({ type: 'end', tileId: id, tileMoves: availableMoves });
    }

    function handleDrop (ev: React.DragEvent<HTMLDivElement>) {
        ev.preventDefault();
        return false;
    }

    const attributes = {
        title: content,
        style: { '--color': getPlayerColor(playerId) },
        className: 'hex tile',
        draggable: !!availableMoves.length,
        onDragStart: handleDragStart,
        onDragEnd: handleDragEnd,
        onDrop: handleDrop
    };

    return <div {...attributes}><span>{content}</span></div>;
}

Tile.displayName = 'Tile';
Tile.defaultProps = defaultProps;

React.memo(Tile, ((p, n) => p.id == n.id && p.availableMoves.length == n.availableMoves.length));

export default Tile;