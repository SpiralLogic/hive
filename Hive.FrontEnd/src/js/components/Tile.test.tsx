import * as React from 'react';
import {render, unmountComponentAtNode} from 'react-dom';
import {act, Simulate} from 'react-dom/test-utils';
import {create} from 'react-test-renderer';
import {Tile} from './Tile';
import {GameContext, HiveContext} from "../gameContext";

const mockContext = (): GameContext => ({
    getPlayerColor: jest.fn().mockReturnValueOnce('pink').mockReturnValueOnce('sky'),
    moveTile: jest.fn(),
    gameState: jest.genMockFromModule('../domain/gameState'),
});

const player1Tile = { id: 1, playerId: 1, content: 'ant', availableMoves: [{ q: 1, r: 1 }] };
const player2Tile = { id: 2, playerId: 2, content: 'fly', availableMoves: [] };

const tileJSX = () =>
    <HiveContext.Provider value={mockContext()}>
        <Tile key="1" {...player1Tile}/>
        <Tile key="2" {...player2Tile}/>
    </HiveContext.Provider>;

let container: HTMLDivElement;

beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
    act(() => { render(tileJSX(), container);});
});

describe('Tile Render', () => {
    test('tile color is the player\'s color', () => {
        const tiles = document.querySelectorAll<HTMLDivElement>('.tile');
        expect(tiles[0].style.getPropertyValue('--color')).toBe('pink');
        expect(tiles[1].style.getPropertyValue('--color')).toBe('sky');
    });

    test('tile has content', () => {
        const tiles = document.querySelectorAll<HTMLDivElement>('.tile');
        expect(tiles[0].textContent).toBe('ant');
        expect(tiles[1].textContent).toBe('fly');
    });
});

describe('Tile drag and drop', () => {
    const attachCellToDom = (q: string, r: string) => {
        const cell = document.createElement('div');
        cell.classList.add('cell');
        cell.dataset['coords'] = `${q},${r}`;
        document.body.appendChild(cell);
        return cell;
    };

    test('Tile is draggable when there are available moves', () => {
        const tiles = document.querySelectorAll<HTMLDivElement>('.tile');
        expect(tiles[0].attributes.getNamedItem('draggable')).toHaveProperty('value', 'true');
    });

    test('is *not* draggable when there are no moves available', () => {
        const tiles = document.querySelectorAll<HTMLDivElement>('.tile');
        expect(tiles[1].attributes.getNamedItem('draggable')).toHaveProperty('value', 'false');
    });

    test('on drag start available cells are given \'valid-cell\' class', () => {
        const availableCell = attachCellToDom('1', '1');
        const tiles = document.querySelectorAll<HTMLDivElement>('.tile');
        const mockDataTransfer = {setData: jest.fn().mockReturnValueOnce(2)};

        // @ts-ignore
        Simulate.dragStart(tiles[0], {dataTransfer: mockDataTransfer});

        expect(availableCell.classList).toContain('valid-cell');
    });

    test('on dragEnd cells have drag class removed', () => {
        const cell = attachCellToDom('1', '1');
        cell.classList.add('active', 'valid-cell', 'invalid-cell')

        const tiles = document.querySelectorAll<HTMLDivElement>('.tile');
        Simulate.dragEnd(tiles[0]);

        expect(cell.classList).not.toContain('valid-cell');
        expect(cell.classList).not.toContain('invalid-cell');
        expect(cell.classList).not.toContain('active');
    });

    test('on drag unavailable cells don\'t have \'valid-cell\' class', () => {
        const unavailableCell = attachCellToDom('0', '0');
        const tiles = document.querySelectorAll<HTMLDivElement>('.tile');
        const mockDataTransfer = {setData: jest.fn().mockReturnValueOnce(2)};

        // @ts-ignore
        Simulate.dragStart(tiles[0], {dataTransfer: mockDataTransfer});

        expect(unavailableCell.classList).not.toContain('valid-cell');
    });
});

describe('Tile Snapshot', () => {
    test('matches current snapshot', () => {
        const component = create(tileJSX());

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});

afterEach(() => {
    unmountComponentAtNode(container);
    container.remove();
});