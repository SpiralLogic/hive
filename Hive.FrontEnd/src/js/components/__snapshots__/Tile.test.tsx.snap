// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tile Snapshot 1`] = `
<div
  className="hex tile"
  style={
    Object {
      "--color": "pink",
    }
  }
  title="ant"
>
  <span>
    ant
  </span>
</div>
`;

exports[`Tile component matches current snapshot 1`] = `
<div
  className="hex tile"
  style={
    Object {
      "--color": undefined,
    }
  }
  title="ant"
>
  <span>
    ant
  </span>
</div>
`;

exports[`Tile has players color 1`] = `null`;
